server/src/main/java/org/opensearch/gateway/PrimaryShardAllocator.java

```diff
@@ -483,13 +535,162 @@ static class NodesToAllocate {
+        // The replication checkpoint of the shard
+        private final ReplicationCheckpoint replicationCheckpoint;
+        // The DiscoveryNode the shard belongs to
+        private final DiscoveryNode node;
+
+        /**
+         * Constructs a new NodeShardState with the given parameters.
+         * @param node The DiscoveryNode the shard belongs to.
+         * @param allocationId The allocation ID of the shard.
+         * @param primary Whether the shard is a primary shard.
+         * @param replicationCheckpoint The replication checkpoint of the shard.
+         * @param storeException Any store exception associated with the shard.
+         */
+        public NodeShardState(
+            DiscoveryNode node,
+            String allocationId,
+            boolean primary,
+            ReplicationCheckpoint replicationCheckpoint,
+            Exception storeException
+        ) {
+            this.node = node;
+            this.allocationId = allocationId;
+            this.primary = primary;
+            this.replicationCheckpoint = replicationCheckpoint;
+            this.storeException = storeException;
+        }
+
+        /**
+         * Returns the allocation ID of the shard.
+         * @return The allocation ID of the shard.
+         */
+        public String allocationId() {
+            return this.allocationId;
+        }
+
+        /**
+         * Returns whether the shard is a primary shard.
+         * @return True if the shard is a primary shard, false otherwise.
+         */
+        public boolean primary() {
+            return this.primary;
+        }
+
+        /**
+         * Returns the replication checkpoint of the shard.
+         * @return The replication checkpoint of the shard.
+         */
+        public ReplicationCheckpoint replicationCheckpoint() {
+            return this.replicationCheckpoint;
+        }
+
+        /**
+         * Returns any store exception associated with the shard.
+         * @return The store exception associated with the shard, or null if there isn't one.
+         */
+        public Exception storeException() {
+            return this.storeException;
+        }
+
+        /**
+         * Returns the DiscoveryNode the shard belongs to.
+         * @return The DiscoveryNode the shard belongs to.
+         */
+        public DiscoveryNode getNode() {
+            return this.node;
+        }
+    }
+
+    /**
+     * The NodeShardStates class manages pairs of {@link NodeShardState} and {@link DiscoveryNode}.
+     * It uses a TreeMap to ensure that the entries are sorted based on the natural
+     * ordering of the {@link NodeShardState} keys, or according to a provided Comparator.
+     * <p>
+     * The TreeMap is implemented using a Red-Black tree, which provides efficient
+     * performance for common operations such as adding, removing, and retrieving
+     * elements.
+     * @see TreeMap
+     */
+    protected static class NodeShardStates {
```

Comment:
Why do we need this class when all we are doing is delegating method calls to TreeMap. Also, want to understand the need of TreeMap in the first place. You already have a DiscoveryNode within NodeShardState then why do you need an additional mapping?
I can see that you only need allocationID to build matchingAllocationFirst map. If that's the only use case then why not create a simple List and just keep adding comparators and then sort in the end just the way it was before.

Reply:
We have created this class to basically provide an abstracted out NodeShardStates object which encapsulates shard states for each of the node, and can be used to pass in methods rather than using directly List or TreeMap.
We can use a List in currently, but when we will implement the BatchAllocator, we will have to Override this class to use a Map only, because we have to maintain all the fetchCall made to nodes and shard present on those nodes.
My earlier implementation had a interface in BaseGatewayShard, which was implemented in PSA using a TreeSet to maintain this list in a sorted order. That interface could than be used in BatchAllocator to implement using the TreeMap. Later @amkhar suggested to keep this class simple by just using the TreeMap that can be used in both current Allocator and upcoming batchAllocator.

Reply:
I meant the following earlier 

```java
public class NodeShardStates {
    private List<NodeShardState> shardStates;
    private DiscoveryNode discoveryNode;
}
``` 
And pass around the instance of this. Any concern in using this?
This will require a very minimal change and PR will get trimmed significantly.

Reply:
Thanks for the suggestion Vikas. We were using this TreeMap earlier because we did not have a encapsulation of the data. Actually, what you have shared will work now. I will modify my code to use List only in next commit.

Reply:
Using List of NodeShardState in the latest commit